lang/py/src/avro/codecs.py

```diff
@@ -0,0 +1,181 @@
+  has_zstandard = True
+except ImportError:
+  has_zstandard = False
+
+
+class Codec(object):
+  def compress(self, data):
+    raise NotImplementedError()
+
+  def decompress(self, readers_decoder):
+    raise NotImplementedError()
+
+
+class NullCodec(Codec):
+  def compress(self, data):
+    return data, len(data)
+
+  def decompress(self, readers_decoder):
+    readers_decoder.skip_long()
+    return readers_decoder
+
+
+class DeflateCodec(Codec):
+  def compress(self, data):
+    # The first two characters and last character are zlib
+    # wrappers around deflate data.
+    compressed_data = zlib.compress(data)[2:-1]
+    return compressed_data, len(compressed_data)
+
+  def decompress(self, readers_decoder):
+    # Compressed data is stored as (length, data), which
+    # corresponds to how the "bytes" type is encoded.
+    data = readers_decoder.read_bytes()
+    # -15 is the log of the window size; negative indicates
+    # "raw" (no zlib headers) decompression.  See zlib.h.
+    uncompressed = zlib.decompress(data, -15)
+    return io.BinaryDecoder(StringIO(uncompressed))
+
+
+class BZip2Codec(Codec):
+  def compress(self, data):
+    compressed_data = bz2.compress(data)
+    return compressed_data, len(compressed_data)
+
+  def decompress(self, readers_decoder):
+    length = readers_decoder.read_long()
+    data = readers_decoder.read(length)
+    uncompressed = bz2.decompress(data)
+    return io.BinaryDecoder(StringIO(uncompressed))
+
+
+class SnappyCodec(Codec):
+  def compress(self, data):
+    compressed_data = snappy.compress(data)
+    # A 4-byte, big-endian CRC32 checksum
+    compressed_data += STRUCT_CRC32.pack(crc32(data) & 0xffffffff)
+    return compressed_data, len(compressed_data)
+
+  def decompress(self, readers_decoder):
+    # Compressed data includes a 4-byte CRC32 checksum
+    length = readers_decoder.read_long()
+    data = readers_decoder.read(length - 4)
+    uncompressed = snappy.decompress(data)
+    checksum = readers_decoder.read(4)
+    self.check_crc32(uncompressed, checksum)
+    return io.BinaryDecoder(StringIO(uncompressed))
+
+  def check_crc32(self, bytes, checksum):
+    checksum = STRUCT_CRC32.unpack(checksum)[0];
+    if crc32(bytes) & 0xffffffff != checksum:
+      raise schema.AvroException("Checksum failure")
+
+
+class ZstandardCodec(Codec):
+  def compress(self, data):
+    compressed_data = zstd.ZstdCompressor().compress(data)
+    return compressed_data, len(compressed_data)
+
+  def decompress(self, readers_decoder):
+    length = readers_decoder.read_long()
+    data = readers_decoder.read(length)
+    uncompressed = bytearray()
+    dctx = zstd.ZstdDecompressor()
+    with dctx.stream_reader(StringIO(data)) as reader:
+      while True:
+        chunk = reader.read(16384)
+        if not chunk:
+          break
+        uncompressed.extend(chunk)
+    return io.BinaryDecoder(StringIO(uncompressed))
+
+
+class Codecs(object):
+  @staticmethod
+  def get_codec(codec_name):
```

Comment:
This may be too complex a factory for our needs. Another approach might be to assign a name to each codec class, e.g.

```python
class SnappyCodec(Codec):
    name = "snappy"
    ...
```

and then use the codec classes instead of the string names in the implementation, like

```python
def __init__(self, writer, datum_writer, writers_schema=None, codec=avro.codecs.NullCodec):
    ...
    self.set_meta(CODEC_KEY, codec.name)
    ...
```

I believe this is simpler than maintaining the explicit mapping of string-to-codec-class in multiple places.